Everything's almost working. There's one last thing left to solve.

When building, CMake complains about missing XWayland include paths:
```c
hyprland> CMake Error in CMakeLists.txt:
hyprland>   Imported target "PkgConfig::xdeps" includes non-existent path
hyprland>     "/nix/store/7q83znq247cfc65h3zzq6l82z8dr9d56-xwayland-23.2.6/include"
hyprland>   in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:
hyprland>   * The path was deleted, renamed, or moved to another location.
hyprland>   * An install or uninstall procedure did not complete successfully.
hyprland>   * The installation package was faulty and references files it does not
hyprland>   provide.
```

Checking the out dir of `xwayland`, it seems there isn't any `include` dir, even though it's referenced by its pkg-config file.
Upstream, we fixed this by using [an override](https://github.com/hyprwm/Hyprland/blob/47874f09f4d14703fe0c483a46b345c7be601ace/nix/overlays.nix#L67-L75).

Ideally, this should be fixed in xwayland upstream. I've tested the latest tag, `24.0.99.902`, but it still references the non-existent path. Pinging @emantor.

Alternatively, we could solve the problem if we could remove the path from `INTERFACE_INCLUDE_DIRECTORIES`.

--- a/pkgs/applications/window-managers/hyprwm/hyprland/default.nix
+++ b/pkgs/applications/window-managers/hyprwm/hyprland/default.nix
@@ -8,27 +8,29 @@
 , ninja
 , binutils
 , cairo
+, epoll-shim
 , git
 , hyprcursor
 , hyprland-protocols
 , hyprlang
+, hyprwayland-scanner
 , jq
 , libGL
 , libdrm
 , libexecinfo
 , libinput
-, libxcb
+, libuuid
 , libxkbcommon
 , mesa
 , pango
 , pciutils
+, pkgconf
 , python3
 , systemd
 , tomlplusplus
 , wayland
 , wayland-protocols
 , wayland-scanner
-, xcbutilwm
 , xwayland
 , hwdata
 , seatd
@@ -51,29 +53,27 @@ assert lib.assertMsg (!hidpiXWayland) "The option `hidpiXWayland` has been remov
 
 stdenv.mkDerivation (finalAttrs: {
   pname = "hyprland" + lib.optionalString debug "-debug";
-  version = "0.39.1";
+  version = "0.40.0-unstable-2024-05-12";
+
   src = fetchFromGitHub {
     owner = "hyprwm";
     repo = finalAttrs.pname;
     fetchSubmodules = true;
-    rev = "v${finalAttrs.version}";
-    hash = "sha256-7L5rqQRYH2iyyP5g3IdXJSlATfgnKhuYMf65E48MVKw=";
+    rev = "2ccd45a84475fab46c6fecd2fe226d3173104743";
+    hash = "sha256-nBCQuRl4sS/G/OUS+txeelFShBEgSk2OrN6kBYMHuOg=";
   };
 
   postPatch = ''
     # Fix hardcoded paths to /usr installation
     sed -i "s#/usr#$out#" src/render/OpenGL.cpp
 
-    # Generate version.h
-    cp src/version.h.in src/version.h
-    substituteInPlace src/version.h \
-      --replace-fail "@HASH@" '${finalAttrs.src.rev}' \
-      --replace-fail "@BRANCH@" "" \
-      --replace-fail "@MESSAGE@" "" \
-      --replace-fail "@DATE@" "2024-04-16" \
-      --replace-fail "@TAG@" "" \
-      --replace-fail "@DIRTY@" ""
+    # Remove extra @PREFIX@ to fix pkg-config paths
+    sed -i "s#@PREFIX@/##g" hyprland.pc.in
   '';
+
+  # used by version.sh
+  DATE = "2024-05-12";
+  HASH = finalAttrs.src.rev;
 
   depsBuildBuild = [
     # to find wayland-scanner when cross-compiling
@@ -82,14 +82,15 @@ stdenv.mkDerivation (finalAttrs: {
 
   nativeBuildInputs = [
     hwdata
+    hyprwayland-scanner
     jq
     makeWrapper
-    meson
+    cmake
+    meson # for wlroots
     ninja
     pkg-config
     wayland-scanner
-    cmake # for subproject udis86
-    python3
+    python3 # for udis86
   ];
 
   outputs = [
@@ -107,6 +108,7 @@ stdenv.mkDerivation (finalAttrs: {
     libGL
     libdrm
     libinput
+    libuuid
     libxkbcommon
     mesa
     wayland
@@ -121,27 +123,33 @@ stdenv.mkDerivation (finalAttrs: {
     xorg.xcbutilerrors
     xorg.xcbutilrenderutil
   ]
+  ++ lib.optionals stdenv.hostPlatform.isBSD [ epoll-shim ]
   ++ lib.optionals stdenv.hostPlatform.isMusl [ libexecinfo ]
-  ++ lib.optionals enableXWayland [ libxcb xcbutilwm xwayland ]
+  ++ lib.optionals enableXWayland [
+    xorg.libxcb
+    xorg.libXdmcp
+    xorg.xcbutil
+    xorg.xcbutilwm
+    xwayland
+  ]
   ++ lib.optionals withSystemd [ systemd ];
 
-  mesonBuildType =
+  cmakeBuildType =
     if debug
-    then "debug"
-    else "release";
+    then "Debug"
+    else "RelWithDebInfo";
 
-  mesonAutoFeatures = "enabled";
 
-  mesonFlags = [
-    (lib.mesonEnable "xwayland" enableXWayland)
-    (lib.mesonEnable "legacy_renderer" legacyRenderer)
-    (lib.mesonEnable "systemd" withSystemd)
+  cmakeFlags = [
+    (lib.cmakeBool "NO_XWAYLAND" (!enableXWayland))
+    (lib.cmakeBool "LEGACY_RENDERER" legacyRenderer)
+    (lib.cmakeBool "NO_SYSTEMD" (!withSystemd))
   ];
 
   postInstall = ''
     ${lib.optionalString wrapRuntimeDeps ''
       wrapProgram $out/bin/Hyprland \
-        --suffix PATH : ${lib.makeBinPath [binutils pciutils stdenv.cc]}
+        --suffix PATH : ${lib.makeBinPath [binutils pciutils pkgconf]}
     ''}
   '';
 

--- a/pkgs/by-name/hy/hyprwayland-scanner/package.nix
+++ b/pkgs/by-name/hy/hyprwayland-scanner/package.nix
@@ -1,0 +1,39 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, pkg-config
+, pugixml
+,
+}:
+stdenv.mkDerivation (finalAttrs: {
+  pname = "hyprwayland-scanner";
+  version = "0.3.7";
+
+  src = fetchFromGitHub {
+    owner = "hyprwm";
+    repo = "hyprwayland-scanner";
+    rev = "v${finalAttrs.version}";
+    hash = "sha256-BPc9FdIzQAmP1JnYGfr+3O30Yp1BywUvKUz57ZOZ39A=";
+  };
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+  ];
+
+  buildInputs = [
+    pugixml
+  ];
+
+  doCheck = true;
+
+  meta = with lib; {
+    homepage = "https://github.com/hyprwm/hyprwayland-scanner";
+    description = "A Hyprland version of wayland-scanner in and for C++";
+    license = licenses.bsd3;
+    maintainers = with maintainers; [ fufexan ];
+    mainProgram = "hyprwayland-scanner";
+    platforms = platforms.linux;
+  };
+})

